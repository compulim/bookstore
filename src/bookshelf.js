import { blob } from 'azure-storage-fs';
import { posix } from 'path';

import onErrorResumeNext from './onErrorResumeNext';
import promisify from './promisify';
import removeKey from './removeKey';

const JSON_CONTENT_SETTINGS_OPTIONS = {
  contentSettings: {
    contentType: 'application/json'
  }
};

export default function bookshelf(blobService, container, prefix = '') {
  function coverBlob(id) {
    return `${ prefix }${ id }.json`;
  }

  function chapterBlob(id, sectionName) {
    return posix.join(`${ prefix }${ id }`, `${ sectionName }.json`);
  }
  async function createBlobFS() {
    await promisify(blobService.createContainerIfNotExists, blobService)(container);

    return {
      ...blob(
        blobService,
        container
      ).promise,
      blobService,
      container
    };
  }

  async function createBook(id, cover, chapters) {
    const fs = await createBlobFS();

    try {
      await Promise.all(Object.keys(chapters).map(chapterName => {
        return fs.writeFile(
          chapterBlob(id, chapterName),
          JSON.stringify(chapters[chapterName]),
          JSON_CONTENT_SETTINGS_OPTIONS
        );
      }));

      await fs.writeFile(coverBlob(id), '', { metadata: { ...cover, __id: id } });
    } catch (err) {
      await onErrorResumeNext(() => Promise.all(
        Object.keys(chapters).map(chapterName => fs.unlink(chapterBlob(id, chapterName)))
      ));

      throw err;
    }

    return { ...chapters, cover, id };
  }

  async function readBookCover(fs, id) {
    const { metadata } = await fs.stat(coverBlob(id), { metadata: true });

    return removeKey(metadata, '__id');
  }

  async function updateBookCover(id, updater) {
    const fs = await createBlobFS();
    const cover = await readBookCover(fs, id);
    const nextCover = updater(cover);

    await fs.setMetadata(coverBlob(id), { ...nextCover, __id: id });

    return { id, cover: nextCover };
  }

  async function listBook() {
    const fs = await createBlobFS();
    const { entries } = await promisify(fs.blobService.listBlobsSegmentedWithPrefix, fs.blobService)(fs.container, prefix, null, { delimiter: '/', include: 'metadata' });

    return entries.reduce((books, entry) => ({
      ...books,
      [entry.metadata.__id]: {
        id: entry.metadata.__id,
        cover: removeKey(entry.metadata, '__id')
      }
    }), {});
  }

  async function deleteBook(id) {
    const fs = await createBlobFS();
    const { entries } = await promisify(fs.blobService.listBlobsSegmentedWithPrefix, fs.blobService)(fs.container, `${ prefix }${ id }/`, null, { delimiter: '/' });

    await fs.unlink(coverBlob(id));
    await Promise.all(entries.map(({ name }) => fs.unlink(name)));
  }

  async function readChapter(id, chapterName) {
    const fs = await createBlobFS();

    return JSON.parse(await fs.readFile(chapterBlob(id, chapterName), { encoding: 'utf8' }));
  }

  async function updateChapter(id, chapterName, updater) {
    const fs = await createBlobFS();
    const cover = await readBookCover(fs, id);
    const chapter = await readChapter(id, chapterName);
    const { chapter: nextChapter, cover: nextCover } = updater({ chapter, cover });

    await fs.writeFile(chapterBlob(id, chapterName), JSON.stringify(nextChapter), JSON_CONTENT_SETTINGS_OPTIONS);

    if (nextCover) {
      try {
        await fs.setMetadata(coverBlob(id), { ...nextCover, __id: id });
      } catch (err) {
        await onErrorResumeNext(() => fs.writeFile(chapterBlob(id, chapterName), JSON.stringify(chapter), JSON_CONTENT_SETTINGS_OPTIONS));
        throw err;
      }
    }

    return {
      [chapterName]: nextChapter,
      id,
      cover: nextCover || cover
    };
  }

  return {
    createBook,
    updateBookCover,
    listBook,
    deleteBook,
    readChapter,
    updateChapter
  };
}
