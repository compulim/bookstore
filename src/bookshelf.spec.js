import { blobService } from './bookshelf.mock';
import mockFS from 'azure-storage-fs';
import Bookshelf from './bookshelf';

let bookshelf;
let createBookResult;

beforeEach(async () => {
  blobService.files = {};
  blobService.createContainerIfNotExists.mock.calls = [];

  bookshelf = Bookshelf(blobService, 'user-1', 'travel/');

  createBookResult = await bookshelf.createBook(
    '1',
    {
      name: 'Japan'
    },
    {
      members: ['user-2', 'user-3'],
      poi: ['SkyTree']
    }
  );

  expect(blobService.createContainerIfNotExists).toHaveBeenCalledTimes(1);
  expect(blobService.createContainerIfNotExists.mock.calls[0][0]).toBe('user-1');
});

test('create book', async () => {
  expect(createBookResult).toMatchSnapshot();
  expect(blobService.getFiles()).toMatchSnapshot();
});

test('update book cover', async () => {
  const result = await bookshelf.updateBookCover('1', cover => ({
    ...cover,
    owner: 'John Doe'
  }));

  expect(result).toMatchSnapshot();
  expect(blobService.getFiles()).toMatchSnapshot();
});

test('list book', async () => {
  await bookshelf.createBook('2', { name: 'Taipei' }, {});

  const result = await bookshelf.listBook();

  expect(result).toMatchSnapshot();
});

test('delete book', async () => {
  expect(blobService.getFiles()).toMatchSnapshot();

  await bookshelf.deleteBook('1');

  expect(blobService.getFiles()).toMatchSnapshot();
});

test('read chapter', async () => {
  const chapter = await bookshelf.readChapter('1', 'poi');

  expect(chapter).toMatchSnapshot();
});

test('update chapter with cover', async () => {
  const updateResult = await bookshelf.updateChapter('1', 'poi', ({ chapter, cover }) => {
    expect(cover).toMatchSnapshot();
    expect(chapter).toMatchSnapshot();

    return {
      chapter: [...chapter, 'Asakusa'],
      cover: { ...cover, owner: 'John Doe' }
    };
  });

  expect(updateResult).toMatchSnapshot();
  expect(blobService.getFiles()).toMatchSnapshot();
});

test('update chapter without cover', async () => {
  const updateResult = await bookshelf.updateChapter('1', 'poi', ({ chapter }) => {
    expect(chapter).toMatchSnapshot();

    return { chapter: [...chapter, 'Asakusa'] };
  });

  expect(updateResult).toMatchSnapshot();
  expect(blobService.getFiles()).toMatchSnapshot();
});
