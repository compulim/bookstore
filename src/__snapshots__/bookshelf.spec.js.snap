// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create book 1`] = `
Object {
  "cover": Object {
    "name": "Japan",
  },
  "id": "1",
  "members": Array [
    "user-2",
    "user-3",
  ],
  "poi": Array [
    "SkyTree",
  ],
}
`;

exports[`create book 2`] = `
Object {
  "travel/1.json": Object {
    "data": "",
    "metadata": Object {
      "__id": "1",
      "name": "Japan",
    },
  },
  "travel/1/members.json": Object {
    "data": "[\\"user-2\\",\\"user-3\\"]",
    "metadata": Object {},
  },
  "travel/1/poi.json": Object {
    "data": "[\\"SkyTree\\"]",
    "metadata": Object {},
  },
}
`;

exports[`delete book 1`] = `
Object {
  "travel/1.json": Object {
    "data": "",
    "metadata": Object {
      "__id": "1",
      "name": "Japan",
    },
  },
  "travel/1/members.json": Object {
    "data": "[\\"user-2\\",\\"user-3\\"]",
    "metadata": Object {},
  },
  "travel/1/poi.json": Object {
    "data": "[\\"SkyTree\\"]",
    "metadata": Object {},
  },
}
`;

exports[`delete book 2`] = `Object {}`;

exports[`list book 1`] = `
Object {
  "1": Object {
    "cover": Object {
      "name": "Japan",
    },
    "id": "1",
  },
  "2": Object {
    "cover": Object {
      "name": "Taipei",
    },
    "id": "2",
  },
}
`;

exports[`read chapter 1`] = `
Array [
  "SkyTree",
]
`;

exports[`update book cover 1`] = `
Object {
  "cover": Object {
    "name": "Japan",
    "owner": "John Doe",
  },
  "id": "1",
}
`;

exports[`update book cover 2`] = `
Object {
  "travel/1.json": Object {
    "data": "",
    "metadata": Object {
      "__id": "1",
      "name": "Japan",
      "owner": "John Doe",
    },
  },
  "travel/1/members.json": Object {
    "data": "[\\"user-2\\",\\"user-3\\"]",
    "metadata": Object {},
  },
  "travel/1/poi.json": Object {
    "data": "[\\"SkyTree\\"]",
    "metadata": Object {},
  },
}
`;

exports[`update chapter with cover 1`] = `
Object {
  "name": "Japan",
}
`;

exports[`update chapter with cover 2`] = `
Array [
  "SkyTree",
]
`;

exports[`update chapter with cover 3`] = `
Object {
  "cover": Object {
    "name": "Japan",
    "owner": "John Doe",
  },
  "id": "1",
  "poi": Array [
    "SkyTree",
    "Asakusa",
  ],
}
`;

exports[`update chapter with cover 4`] = `
Object {
  "travel/1.json": Object {
    "data": "",
    "metadata": Object {
      "__id": "1",
      "name": "Japan",
      "owner": "John Doe",
    },
  },
  "travel/1/members.json": Object {
    "data": "[\\"user-2\\",\\"user-3\\"]",
    "metadata": Object {},
  },
  "travel/1/poi.json": Object {
    "data": "[\\"SkyTree\\",\\"Asakusa\\"]",
    "metadata": Object {},
  },
}
`;

exports[`update chapter without cover 1`] = `
Array [
  "SkyTree",
]
`;

exports[`update chapter without cover 2`] = `
Object {
  "cover": Object {
    "name": "Japan",
  },
  "id": "1",
  "poi": Array [
    "SkyTree",
    "Asakusa",
  ],
}
`;

exports[`update chapter without cover 3`] = `
Object {
  "travel/1.json": Object {
    "data": "",
    "metadata": Object {
      "__id": "1",
      "name": "Japan",
    },
  },
  "travel/1/members.json": Object {
    "data": "[\\"user-2\\",\\"user-3\\"]",
    "metadata": Object {},
  },
  "travel/1/poi.json": Object {
    "data": "[\\"SkyTree\\",\\"Asakusa\\"]",
    "metadata": Object {},
  },
}
`;
